# Ensembles:

In [1]:
import warnings
warnings.filterwarnings('ignore')
import numpy as np
import pandas as pd
from pathlib import Path
from collections import Counter
from sklearn.metrics import balanced_accuracy_score, confusion_matrix
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, StandardScaler
from imblearn.metrics import classification_report_imbalanced
from imblearn.ensemble import BalancedRandomForestClassifier, EasyEnsembleClassifier

C:\Users\bgregory\anaconda3\envs\pyvizenv\lib\site-packages\ipykernel\parentpoller.py:113: UserWarning: Parent poll failed.  If the frontend dies,
                the kernel may be left running.  Please let us know
                about your system (bitness, Python, etc.) at
                ipython-dev@scipy.org
  warnings.warn("""Parent poll failed.  If the frontend dies,


In [ ]:
# Load csv to DataFrame:
file_path = Path('Resources/LoanStats_2019Q1.csv')
stats_df = pd.read_csv(file_path)
stats_df.head()

In [ ]:
# Create LabelEncoder instance:
le = LabelEncoder()

# Encode home_ownership column:
le.fit(stats_df['home_ownership'])
stats_df['home_ownership'] = le.transform(stats_df['home_ownership'])

# Encode verification_status column:
le.fit(stats_df['verification_status'])
stats_df['verification_status'] = le.transform(stats_df['verification_status'])

# Encode issue_d column:
le.fit(stats_df['issue_d'])
stats_df['issue_d'] = le.transform(stats_df['issue_d'])

# Encode pymnt_plan column:
le.fit(stats_df["pymnt_plan"])
stats_df['pymnt_plan'] = le.transform(stats_df['pymnt_plan'])

# Encode loan_status column:
le.fit(stats_df['loan_status'])
stats_df['loan_status'] = le.transform(stats_df['loan_status'])

# Encode next_pymnt_d column:
le.fit(stats_df['next_pymnt_d'])
stats_df['next_pymnt_d'] = le.transform(stats_df['next_pymnt_d'])

# Encode application_type column:
le.fit(stats_df['application_type'])
stats_df['application_type'] = le.transform(stats_df['application_type'])


# Encode initial_list_status column:
le.fit(stats_df['initial_list_status'])
stats_df['initial_list_status'] = le.transform(stats_df['initial_list_status'])

# Encode hardship_flag column:
le.fit(stats_df['hardship_flag'])
stats_df['hardship_flag'] = le.transform(stats_df['hardship_flag'])

# Encode debt_settlement_flag column:
le.fit(stats_df['debt_settlement_flag'])
stats_df['debt_settlement_flag'] = le.transform(stats_df['debt_settlement_flag'])

stats_df.head()

## Split the Data into Training and Testing

In [4]:
# Define target:
y = stats_df['loan_status']

# Check the balance of target values:
y.value_counts()

1    68470
0      347
Name: loan_status, dtype: int64

In [5]:
# Define features:
x = stats_df.copy()
x.drop('loan_status',axis=1,inplace=True)
x.describe()

loan_amnt      int_rate   installment  home_ownership    annual_inc  \
count  68817.000000  68817.000000  68817.000000    68817.000000  6.881700e+04   
mean   16677.594562      0.127718    480.652863        1.812779  8.821371e+04   
std    10277.348590      0.048130    288.062432        0.941313  1.155800e+05   
min     1000.000000      0.060000     30.890000        0.000000  4.000000e+01   
25%     9000.000000      0.088100    265.730000        1.000000  5.000000e+04   
50%    15000.000000      0.118000    404.560000        1.000000  7.300000e+04   
75%    24000.000000      0.155700    648.100000        3.000000  1.040000e+05   
max    40000.000000      0.308400   1676.230000        3.000000  8.797500e+06   

       verification_status       issue_d  pymnt_plan           dti  \
count         68817.000000  68817.000000     68817.0  68817.000000   
mean              0.669994      0.805542         0.0     21.778153   
std               0.719105      0.714932         0.0     20.199244   
min               0.000000      0.000000         0.0      0.000000   
25%               0.000000      0.000000         0.0     13.890000   
50%               1.000000      1.000000         0.0     19.760000   
75%               1.000000      1.000000         0.0     26.660000   
max               2.000000      2.000000         0.0    999.000000   

        delinq_2yrs  ...  pct_tl_nvr_dlq  percent_bc_gt_75  \
count  68817.000000  ...    68817.000000      68817.000000   
mean       0.217766  ...       95.057627         30.626217   
std        0.718367  ...        8.326426         33.631463   
min        0.000000  ...       20.000000          0.000000   
25%        0.000000  ...       93.000000          0.000000   
50%        0.000000  ...      100.000000         20.000000   
75%        0.000000  ...      100.000000         50.000000   
max       18.000000  ...      100.000000        100.000000   

       pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  total_bal_ex_mort  \
count          68817.000000    68817.0     6.881700e+04       6.881700e+04   
mean               0.125972        0.0     2.100332e+05       6.133843e+04   
std                0.336732        0.0     1.928088e+05       5.738798e+04   
min                0.000000        0.0     3.600000e+03       2.350000e+02   
25%                0.000000        0.0     6.697700e+04       2.650300e+04   
50%                0.000000        0.0     1.467100e+05       4.535700e+04   
75%                0.000000        0.0     3.036400e+05       7.657000e+04   
max                4.000000        0.0     3.292782e+06       1.295455e+06   

       total_bc_limit  total_il_high_credit_limit  hardship_flag  \
count    68817.000000                6.881700e+04        68817.0   
mean     29734.128558                5.572240e+04            0.0   
std      26795.394232                5.095845e+04            0.0   
min        100.000000                1.270000e+02            0.0   
25%      11600.000000                2.288000e+04            0.0   
50%      22100.000000                4.200000e+04            0.0   
75%      39300.000000                7.249900e+04            0.0   
max     509400.000000                1.426964e+06            0.0   

       debt_settlement_flag  
count               68817.0  
mean                    0.0  
std                     0.0  
min                     0.0  
25%                     0.0  
50%                     0.0  
75%                     0.0  
max                     0.0  

[8 rows x 85 columns]

In [6]:
# Split x and y into x_train, x_test, y_train, y_test:
x_train, x_test, y_train, y_test = train_test_split(x,y,random_state=78)

## Data Pre-Processing

Scale the training and testing data using `StandardScaler`.

In [7]:
# Create a SandardScaler instance:
scaler = StandardScaler()

# Fit the Standard Scaler:
x_scaler = scaler.fit(x_train)

# Scale training and testing data:
x_train_scaled = x_scaler.transform(x_train)
x_test_scaled = x_scaler.transform(x_test)

## Ensemble Learners

Compare the BalancedRandomForestClassifier and EasyEnsembleClassifier algorithms and determine which results in the best performance then complete the folliowing for each algorithm:

1. Train the model using the training data. 
2. Calculate the Balanced Accuracy Score from sklearn.metrics.
3. Display the Confusion Matrix from sklearn.metrics.
4. Generate an Imbalanced Classification Report from imbalanced-learn.
5. For the Balanced Random Forest Classifier, output the feature importance in descending order (most important to least important) along with the feature score.

### Balanced Random Forest Classifier

In [8]:
# Resample the training data with the BalancedRandomForestClassifier
brf_model = BalancedRandomForestClassifier(n_estimators=100,random_state=1)
brf_model = brf_model.fit(x_train_scaled,y_train)

In [9]:
# Calculate the balanced accuracy score:
brf_predictions = brf_model.predict(x_test_scaled)
brf_ba_score = balanced_accuracy_score(y_test,brf_predictions)
display(brf_ba_score)

0.7518737344353009

In [19]:
# Display the confusion matrix:
brf_cm = confusion_matrix(y_test,brf_predictions)
brf_cm_df = pd.DataFrame(brf_cm,index=['Actual 0','Actual 1'],columns=['Predicted 0','Predicted 1'])
brf_cm_df

Predicted 0  Predicted 1
Actual 0           58           34
Actual 1         2168        14945

In [11]:
# Print the imbalanced classification report:
print(classification_report_imbalanced(y_test,brf_predictions))

                   pre       rec       spe        f1       geo       iba       sup

          0       0.03      0.63      0.87      0.05      0.74      0.54        92
          1       1.00      0.87      0.63      0.93      0.74      0.56     17113

avg / total       0.99      0.87      0.63      0.93      0.74      0.56     17205



In [ ]:
# List features in descending order by importance:
importance_sorted = sorted(zip(brf_model.feature_importances_,x.columns),reverse=True)
display(importance_sorted)

### Easy Ensemble Classifier

In [13]:
# Train the EasyEnsembleClassifier:
eec_model = EasyEnsembleClassifier(n_estimators=100,random_state=1)
eec_model.fit(x_train_scaled,y_train)
y_pred_eec = eec_model.predict(x_test)

In [14]:
# Calculate balanced accuracy score:
eec_ba_score = balanced_accuracy_score(y_test,y_pred_eec)
display(eec_ba_score)

0.48017112594290123

In [15]:
# Display confusion matrix:
eec_cm = confusion_matrix(y_test,y_pred_eec)
eec_cm_df = pd.DataFrame(eec_cm,index=['Actual 0','Actual 1'],columns=['Predicted 0','Predicted 1'])
eec_cm_df

Predicted 0  Predicted 1
Actual 0           31           61
Actual 1         6445        10668

In [16]:
# Print imbalanced classification report:
print(classification_report_imbalanced(y_test,y_pred_eec))

                   pre       rec       spe        f1       geo       iba       sup

          0       0.00      0.34      0.62      0.01      0.46      0.20        92
          1       0.99      0.62      0.34      0.77      0.46      0.22     17113

avg / total       0.99      0.62      0.34      0.76      0.46      0.22     17205



### Final Questions

1. Which model had the best balanced accuracy score?

    The Balanced Random Forest Classifier had the best Balanced Accuracy Score by about 27%

2. Which model had the best recall score?

    The Balanced Random Forest Classifier had the best Recall Score by 15%

3. Which model had the best geometric mean score?

    YOUR ANSWER HERE.

4. What are the top three features?

    YOUR ANSWER HERE.